clang/lib/CodeGen/Targets/RISCV.cpp

```diff
@@ -429,9 +429,43 @@ ABIArgInfo RISCVABIInfo::classifyArgumentType(QualType Ty, bool IsFixed,
     return ABIArgInfo::getIgnore();
 
   bool IsSingleCapRecord = false;
-  if (auto *RT = Ty->getAs<RecordType>())
+  bool FitsInTwoRegs = false;
+  auto *RD = dyn_cast_if_present<RecordDecl>(Ty->getAsTagDecl());
+  if (RD) {
     IsSingleCapRecord = Size == getTarget().getCHERICapabilityWidth() &&
-                        getContext().containsCapabilities(RT->getDecl());
+                        getContext().containsCapabilities(RD);
+    FitsInTwoRegs = IsSingleCapRecord;
+
+    StringRef TargetABI = getTarget().getABI();
+    bool IsCheriot = TargetABI == "cheriot" || TargetABI == "cheriot-baremetal";
+
+    // If the target platform is CHERIoT, Check if the record fits in two
+    // registers, that is:
+    // 1. it has two fields
+    // 2. any of the two fields is either a capability or a type whose size can
+    // fit in the data part of a register
+    if (!IsSingleCapRecord && IsCheriot) {
```

Comment:
This check seems a lot more complex than it needs to be.  Why wouldn't this work?

```
   IsTwoCapRecord = Size == 2*getTarget().getCHERICapabilityWidth() &&
                        getContext().containsCapabilities(RD);
```